examples/rust/udp-ping-pong.rs

```diff
@@ -102,240 +103,213 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
-        Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
-    };
-    match libos.bind(sockqd, local) {
-        Ok(()) => (),
-        Err(e) => {
-            // If error, close socket.
-            close(&mut libos, sockqd);
+    pub fn run(&mut self, local: SocketAddrV4, remote: SocketAddrV4, fill_char: u8) -> Result<()> {
+        if let Err(e) = self.libos.bind(self.sockqd, local) {
             anyhow::bail!("bind failed: {:?}", e)
-        },
-    };
-
-    let mut i: usize = 0;
-    loop {
-        // Pop data.
-        let qt: QToken = match libos.pop(sockqd, None) {
-            Ok(qt) => qt,
-            Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("pop failed: {:?}", e)
-            },
-        };
-        let sga: demi_sgarray_t = match libos.wait(qt, None) {
-            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
-            Ok(_) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("unexpected result")
-            },
-            Err(e) => {
-                // If error, close socket.
-                close(&mut libos, sockqd);
-                anyhow::bail!("operation failed: {:?}", e)
-            },
         };
 
-        // Sanity check received data.
-        let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
-        let len: usize = sga.sga_segs[0].sgaseg_len as usize;
-        let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
-        for x in slice {
-            if *x != FILL_CHAR {
-                // If not matching, close socket and free scatter-gather array.
-                freesga(&mut libos, sga);
-                close(&mut libos, sockqd);
-                anyhow::bail!("fill check failed: expected={:?} received={:?}", FILL_CHAR, *x);
+        let mut i: usize = 0;
+        loop {
+            // Pop data.
+            let qt: QToken = match self.libos.pop(self.sockqd, None) {
+                Ok(qt) => qt,
+                Err(e) => anyhow::bail!("pop failed: {:?}", e),
+            };
+            let sga: demi_sgarray_t = match self.libos.wait(qt, None) {
+                Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
+                Ok(_) => anyhow::bail!("unexpected result"),
+                Err(e) => anyhow::bail!("operation failed: {:?}", e),
+            };
+
+            // Sanity check received data.
+            let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+            let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+            for x in slice {
+                if *x != fill_char {
+                    // If not matching, free scatter-gather array.
+                    freesga(&mut self.libos, sga);
```

Comment:
Lets do the same and add Option<demi_sgarray_t> to the server struct, so we can get rid of this line and lines 160, 168 and 173.

Reply:
Done.